src/main/java/com/spotify/dns/AbstractChangeNotifier.java

```diff
@@ -34,18 +36,25 @@
 
   private final AtomicReference<Listener<T>> listenerRef = new AtomicReference<Listener<T>>();
 
+  private final AtomicBoolean listenerNotified = new AtomicBoolean(false);
+
+  private ReentrantLock lock = new ReentrantLock();
```

Comment:
That's a good point, so using a ReentrantLock (in a 'final' field) is probably a good idea!